### PR TITLE
Fix traversal when parent model is loaded

### PR DESCRIPTION
**Pull request by mvorisek**
finish after #855, #860, #861

discussion here: https://github.com/atk4/data/commit/c8cbce3e0ceceab616d990763a6b7f28abadded3#r49718142

fixes https://github.com/mvorisek/atk4-hintable-mirror/blob/1.2.1/tests/Data/HintableModelTest.php#L120

```
$model = new Model\Standard($db);
$model->load(2)->simpleMany->loadAny()->id;
```

must traverse from model with ID = 2 only (before this PR, ID = 2 condition was not applied and it was required to be added manually, now with entity concept we can assume locked ID)

note: this works also with `$model->load{One, Any}()->simpleMany` and also with `Model::duplicate()` which will unset the ID correctly.

### CODE REVIEW

File: src/Persistence/Array_.php [hunk 1]
```diff
@@ -332,7 +332,15 @@ protected function setLimitOrder(Model $model, \Atk4\Data\Action\Iterator $actio
      */
     public function applyScope(Model $model, \Atk4\Data\Action\Iterator $iterator)
     {
-        return $iterator->filter($model->scope());
+        $scope = $model->scope();
+
+        // add entity ID to scope to allow easy traversal
+        if ($model->id_field && $model->getId() !== null) {
```
Comment (mvorisek) [inline src/Persistence/Array_.php:338]: should be probably deduplicated in Model

**Comment by mvorisek** [comment]
@georgehristov I need your feedback on this

File: tests/RandomTest.php [hunk 2]
```diff
@@ -380,7 +380,7 @@ public function testGetTitle()
 
         $m->load(2);
         $this->assertSame('Sue', $m->getTitle()); // loaded returns title_field value
-        $this->assertSame([1 => 'John', 2 => 'Sue'], $m->getTitles()); // all titles
+        $this->assertSame([2 => 'Sue'], $m->getTitles()); // traverse loaded
```
Comment (DarkSide666) [inline tests/RandomTest.php]: I don't think this is correct. `getTitles()` should return names of all records taking applied conditions into account, but it's not related to is model loaded or not. It's like `$m->export([$m->title_field])`.

File: src/Persistence/Sql.php [hunk 3]
```diff
@@ -364,12 +364,23 @@ protected function setLimitOrder(Model $model, Query $query)
     }
 
     /**
-     * Will apply a condition defined inside $condition or $model->scope() onto $query.
+     * Will apply $model->scope() conditions onto $query.
      */
-    public function initQueryConditions(Model $model, Query $query, Model\Scope\AbstractScope $condition = null): void
+    public function initQueryConditions(Model $model, Query $query): void
```
Comment (DarkSide666) [inline src/Persistence/Sql.php:371]: Probably should be `_initQueryConditions` prefixed with underscore

**Review by DarkSide666 — CHANGES_REQUESTED** [review]
Overall I don't think this is correct approach.
Loading record shouldn't add condition to model because that will limit how `getTitles()`, `export()` and other dataset-wise methods will work when model is or is not loaded. That shouldn't happen.

File: src/Persistence/Sql.php (hunk 3 above)
Comment (mvorisek) [reply src/Persistence/Sql.php:371]: why do you think so?

Comment (DarkSide666) [reply src/Persistence/Sql.php:371]: Oh forget about this. I didn't notice other method below.

**Comment by mvorisek** [comment]
All changes reverted and addressed differently, `getTitles()` and other confusing behaviour with entity (loaded model) now thow because of https://github.com/atk4/data/pull/855

File: src/Persistence/Sql.php [hunk 4]
```diff
@@ -369,6 +369,12 @@ protected function setLimitOrder(Model $model, Query $query)
     public function initQueryConditions(Model $model, Query $query): void
     {
         $this->_initQueryConditions($query, $model->scope());
+
+        // add entity ID to scope to allow easy traversal
+        if ($model->id_field && $model->getId() !== null) {
+            $query->group($model->getField($model->id_field));
+            $query->having($model->getField($model->id_field), $model->getId());
```
Comment (mvorisek) [inline src/Persistence/Sql.php:384]: FYI: `having` with `group by` is like `where`, but it does not apply after aggregate functions.

**Review by DarkSide666 — APPROVED** [review]
Approved in advance. Need to fix this approach for MS SQL

File: src/Persistence/Array_.php [hunk 5]
```diff
@@ -185,38 +197,35 @@ public function add(Model $model, array $defaults = []): Model
...
-        if (!isset($this->data[$model->table])) {
-            throw (new Exception('Table was not found in the array data source'))
-                ->addMoreInfo('table', $model->table);
-        }
+        $table = $this->seedDataAndGetTable($model);
 
         if ($id === self::ID_LOAD_ONE || $id === self::ID_LOAD_ANY) {
-            if (count($this->data[$model->table]) === 0) {
+            $action = $this->action($model, 'select');
+            $action->generator->rewind(); // needed for some reasons!
```
Comment (mvorisek) [inline src/Persistence/Array_.php:204]: why is this needed, should this be done in the action itself?
